File: 09_DataStructure_Algorithm/0513_day48/01_Keys and Rooms/Keys and Rooms_dfs.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        int n = rooms.size();

        boolean[] visited = new boolean[n];
        dfs(0, rooms, visited);

        for (boolean v : visited) {
            if (!v) return false;
        }
        return true;
    }

    void dfs(int cur, List<List<Integer>> rooms, boolean[] visited) {
        visited[cur] = true;

        for (int nxt : rooms.get(cur)) {
            if (!visited[nxt]) {
                dfs(nxt, rooms, visited);
            }
        }
    }
}

File: 09_DataStructure_Algorithm/0513_day48/01_Keys and Rooms/Keys and Rooms_dfs.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        n = len(rooms)
        visited = [False] * n

        def dfs(cur):
            visited[cur] = True

            for nxt in rooms[cur]:
                if not visited[nxt]:
                    dfs(nxt)
        
        visited[0] = True
        dfs(0)

        for v in visited:
            if not v:
                return False
        
        return True